src/daft-plan/src/logical_optimization/rules/push_down_projection.rs

```diff
@@ -847,6 +954,79 @@ mod tests {
         Ok(())
     }
 
+    /// Projection<-ActorPoolProject<-ActorPoolProject prunes columns from both ActorPoolProjects
+    #[test]
+    fn test_projection_pushdown_into_double_actorpoolproject() -> DaftResult<()> {
+        use crate::logical_ops::ActorPoolProject;
+        use crate::logical_ops::Project;
+        use common_resource_request::ResourceRequest;
+        use daft_dsl::functions::python::{PythonUDF, StatefulPythonUDF};
+        use daft_dsl::functions::FunctionExpr;
+        use daft_dsl::Expr;
+
+        let scan_op = dummy_scan_operator(vec![
+            Field::new("a", DataType::Int64),
+            Field::new("b", DataType::Boolean),
+            Field::new("c", DataType::Int64),
+        ]);
+        let scan_node = dummy_scan_node(scan_op.clone()).build();
+        let mock_stateful_udf = Expr::Function {
+            func: FunctionExpr::Python(PythonUDF::Stateful(StatefulPythonUDF {
+                name: Arc::new("my-udf".to_string()),
+                num_expressions: 1,
+                return_dtype: DataType::Utf8,
+                resource_request: Some(ResourceRequest::default_cpu()),
+                batch_size: None,
+                concurrency: Some(8),
+            })),
+            inputs: vec![col("a")],
+        }
+        .arced();
+
+        // Select the `udf_results` column, so the ActorPoolProject should apply column pruning to the other columns
+        let plan = LogicalPlan::ActorPoolProject(ActorPoolProject::try_new(
+            scan_node.clone(),
+            vec![col("a"), col("b"), mock_stateful_udf.alias("udf_results_0")],
+        )?)
+        .arced();
+        let plan = LogicalPlan::ActorPoolProject(ActorPoolProject::try_new(
+            plan,
+            vec![
+                col("a"),
+                col("b"),
+                col("udf_results_0"),
+                mock_stateful_udf.alias("udf_results_1"),
+            ],
+        )?)
+        .arced();
+        let plan = LogicalPlan::Project(Project::try_new(
+            plan,
+            vec![col("udf_results_0"), col("udf_results_1")],
```

Comment:
You could also add some aliased columns in here too, since they do not perform computation so you expect them to also be merged with the `ActorPoolProject` correctly

Reply:
Good call-out, added it into the test and verified that the alias gets correctly absorbed into the ActorPoolProject